OpenSim/Simulation/Wrap/WrapCylinder.cpp

```diff
@@ -605,149 +623,106 @@ int WrapCylinder::wrapLine(const SimTK::State& s, SimTK::Vec3& aPoint1, SimTK::V
     // WrapTorus creates a WrapCyl with no connected model, avoid this hack
-    if (!_model.empty() && !getModel().getDisplayHints().isVisualizationEnabled() &&
-            aWrapResult.singleWrap) {
+    if (!_model.empty() &&
+        !getModel().getDisplayHints().isVisualizationEnabled() &&
+        out.singleWrap) {
+
         // Use one WrapSegment/cord instead of finely sampled list of wrap_pt(s)
-        _calc_spiral_wrap_point(
-                r1a, axial_vec, m, ax, sense, 0, theta, wrap_pt);
-        aWrapResult.wrap_pts.append(wrap_pt);
-        _calc_spiral_wrap_point(
-                r1a, axial_vec, m, ax, sense, 1.0, theta, wrap_pt);
-        aWrapResult.wrap_pts.append(wrap_pt);
+        out.wrap_pts.setSize(2);
+        out.wrap_pts[0] = CalcWrapPoint(0);
+        out.wrap_pts[1] = CalcWrapPoint(1.0);
     } 
     else {
+
         // Each muscle segment on the surface of the cylinder should be
         // 0.002 meters long. This assumes the model is in meters, of course.
-        int numWrapSegments = (int)(aWrapResult.wrap_path_length / 0.002);
-        if (numWrapSegments < 1) numWrapSegments = 1;
+        int numWrapSegments = static_cast<int>(out.wrap_path_length / 0.002);
+        if (numWrapSegments < 1) {
+            numWrapSegments = 1;
+        }
+
+        out.wrap_pts.setSize(0);
 
-        for (i = 0; i < numWrapSegments; i++) {
-            double t = (double)i / numWrapSegments;
+        for (int i = 0; i < numWrapSegments; i++) {
+            double t = static_cast<double>(i)/static_cast<double>(numWrapSegments);
 
-            _calc_spiral_wrap_point(
-                    r1a, axial_vec, m, ax, sense, t, theta, wrap_pt);
+            Vec3 wrap_pt = CalcWrapPoint(t);
 
             // adjust r1/r2 tangent points if necessary to achieve tangency with
             // the spiral path:
-            if (i == 1 && iterations < MAX_ITERATIONS &&
-                    !aWrapResult.singleWrap) {
-                bool did_adjust_r2 = false;
-                bool did_adjust_r1 = _adjust_tangent_point(
-                        aPoint1, dn, aWrapResult.r1, wrap_pt);
+            if (i == 1
+                && iterations < MAX_ITERATIONS
+                && !out.singleWrap) {
 
-                SimTK::Vec3 temp_wrap_pt;
+                bool did_adjust_r1 = _adjust_tangent_point(
+                    p1,
+                    g_CylinderDirection,
+                    out.r1,
+                    wrap_pt);
 
-                _calc_spiral_wrap_point(r1a, axial_vec, m, ax, sense, 1.0 - t,
-                        theta, temp_wrap_pt);
+                Vec3 temp_wrap_pt = CalcWrapPoint(1.0 - t);
 
-                did_adjust_r2 = _adjust_tangent_point(
-                        aPoint2, dn, aWrapResult.r2, temp_wrap_pt);
+                bool did_adjust_r2 = _adjust_tangent_point(
+                        p2,
+                        g_CylinderDirection,
+                        out.r2,
+                        temp_wrap_pt);
```

Comment:
https://imgs.xkcd.com/comics/goto.png

Reply:
First time in a long time I've seen a go-to :D

Although on my local branch (which is even more refactored, for a ~15pct boost, by doing stuff like optimized quaternion rotations) I tried to rewrite the code in an iteration loop, which worked fine but (imho) is actually kind of uglier (I hate to admit this) because the original algorithim only logically performs an adjustment if n>=1. It then has to perform two method calls (point adjustment), and this only happens if it's not the final iteration of the loop. Also, the loop block contains the necessary state to finalize the algorithm. This means that you end up with something like a `continue` (jump) on the adjustment step (the `goto`) and a `break` at the bottom of the loop (the fall through), which also felt gross :(